File: cli/src/fp_cli/commands/dev.py
"""fp dev — run locally with hot reload."""

import os
import subprocess
import sys

import typer

from ..console import console, err_console
from ..project import read_project


def dev(
    port: int = typer.Option(8000, "--port", "-p", help="Port to run on"),
):
    """Run the app locally with hot reload."""
    project = read_project()
    entrypoint = project.get("entrypoint", "app.py")

    # Derive module path: app.py -> app:app, src/app.py -> src.app:app
    module = entrypoint.replace("/", ".").removesuffix(".py")
    app_ref = f"{module}:app"

    # Set up environment
    env = os.environ.copy()

    # Inject PLATFORM_MONGO_URI if database is enabled
    if project.get("database"):
        mongo_uri = env.get("PLATFORM_MONGO_URI", "mongodb://localhost:27017")
        env["PLATFORM_MONGO_URI"] = mongo_uri
        console.print(f"  PLATFORM_MONGO_URI={mongo_uri}")

    # Resolve env vars from .fp.yaml
    project_env = project.get("env")
    if project_env and isinstance(project_env, dict):
        for key, value in project_env.items():
            if isinstance(value, str) and value.startswith("${") and value.endswith("}"):
                env_name = value[2:-1]
                env[key] = os.environ.get(env_name, "")
            else:
                env[key] = str(value)

    console.print(f"Running [bold]{app_ref}[/bold] on port {port}")
    console.print("[dim]Ctrl+C to stop[/dim]")
    console.print()

    try:
        subprocess.run(
            [
                sys.executable,
                "-m",
                "uvicorn",
                app_ref,
                "--reload",
                "--host",
                "0.0.0.0",
                "--port",
                str(port),
            ],
            env=env,
        )
    except KeyboardInterrupt:
        pass
    except FileNotFoundError:
        err_console.print(
            "[red]uvicorn not found.[/red] Install it: [bold]pip install uvicorn[/bold]"
        )
        raise typer.Exit(1)

File: cli-go/cmd/dev.go
package cmd

import (
	"fmt"
	"os"
	"os/exec"
	"strings"

	"github.com/spf13/cobra"
	"github.com/thatcatxedo/fastapi-platform/cli-go/internal/project"
)

var devCmd = &cobra.Command{
	Use:   "dev",
	Short: "Run the app locally with hot reload",
	RunE:  runDev,
}

var devPort int

func init() {
	devCmd.Flags().IntVarP(&devPort, "port", "p", 8000, "Port to run on")
}

func runDev(cmd *cobra.Command, args []string) error {
	proj, dir, err := project.ReadFromCwd()
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}

	entrypoint := proj.Entrypoint
	if entrypoint == "" {
		entrypoint = "app.py"
	}

	// app.py -> app:app, src/app.py -> src.app:app
	module := strings.TrimSuffix(entrypoint, ".py")
	module = strings.ReplaceAll(module, "/", ".")
	appRef := module + ":app"

	env := os.Environ()
	if proj.Database != nil {
		mongoURI := os.Getenv("PLATFORM_MONGO_URI")
		if mongoURI == "" {
			mongoURI = "mongodb://localhost:27017"
		}
		env = append(env, "PLATFORM_MONGO_URI="+mongoURI)
		fmt.Printf("  PLATFORM_MONGO_URI=%s\n", mongoURI)
	}

	if proj.Env != nil {
		for k, v := range proj.Env {
			if strings.HasPrefix(v, "${") && strings.HasSuffix(v, "}") {
				v = os.Getenv(v[2 : len(v)-1])
			}
			env = append(env, k+"="+v)
		}
	}

	fmt.Printf("Running %s on port %d\n", appRef, devPort)
	fmt.Println("Ctrl+C to stop")
	fmt.Println()

	uvicorn := exec.Command("python", "-m", "uvicorn", appRef,
		"--reload", "--host", "0.0.0.0", "--port", fmt.Sprintf("%d", devPort))
	uvicorn.Dir = dir
	uvicorn.Env = env
	uvicorn.Stdin = os.Stdin
	uvicorn.Stdout = os.Stdout
	uvicorn.Stderr = os.Stderr

	if err := uvicorn.Run(); err != nil {
		fmt.Fprintln(os.Stderr, "uvicorn not found. Install it: pip install uvicorn")
		os.Exit(1)
	}
	return nil
}
